infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/TopologyDescriptor.java

```diff
@@ -31,80 +31,118 @@ public class TopologyDescriptor {
+    // replication model.
     @Getter
-    private final Map<String, ClusterDescriptor> invalidClusters;
+    private final Map<ClusterDescriptor, Set<LogReplicationMetadata.ReplicationModel>> remoteSourceClusterToReplicationModels;
+
+    //Map of remote clusterId -> ReplicationModels. Contains clusters which will be SINK (w.r.t local cluster) for a
+    // replication model. Used to construct sessions.
+    @Getter
+    private final Map<ClusterDescriptor, Set<LogReplicationMetadata.ReplicationModel>> remoteSinkClusterToReplicationModels;
+
+    // Contains the complete view of topology.
+    @Getter
+    private final Map<String, ClusterConfigurationMsg> allClusterMsgsInTopology = new HashMap<>();
+
 
     /**
-     * Constructor
+     * Constructor used by discoveryService and tests
      *
      * @param topologyMessage proto definition of the topology
+     * @param remoteSinkToReplicationModel remote sink to local cluster and corresponding replication model
+     * @param remoteSourceToReplicationModel remote source to local cluster and corresponding replication model
      */
-    public TopologyDescriptor(TopologyConfigurationMsg topologyMessage) {
+    public TopologyDescriptor(TopologyConfigurationMsg topologyMessage,
+                              Map<ClusterConfigurationMsg, Set<LogReplicationMetadata.ReplicationModel>> remoteSinkToReplicationModel,
+                              Map<ClusterConfigurationMsg, Set<LogReplicationMetadata.ReplicationModel>> remoteSourceToReplicationModel) {
         this.topologyConfigId = topologyMessage.getTopologyConfigID();
-        this.sinkClusters = new HashMap<>();
-        this.sourceClusters = new HashMap<>();
-        this.invalidClusters = new HashMap<>();
-
-        for (ClusterConfigurationMsg clusterConfig : topologyMessage.getClustersList()) {
-            ClusterDescriptor cluster = new ClusterDescriptor(clusterConfig);
-            if (clusterConfig.getRole() == ClusterRole.SOURCE) {
-                sourceClusters.put(cluster.getClusterId(), cluster);
-            } else if (clusterConfig.getRole() == ClusterRole.SINK) {
-                addSinkCluster(cluster);
-            } else {
-                invalidClusters.put(cluster.getClusterId(), cluster);
-            }
-        }
-    }
+        this.remoteSinkClusters = new HashMap<>();
+        this.remoteSinkClusterToReplicationModels = new HashMap<>();
 
-    /**
-     * Constructor
-     *
-     * @param topologyConfigId topology configuration identifier (epoch)
-     * @param sourceCluster source cluster
-     * @param sinkClusters sink cluster's
-     */
-    public TopologyDescriptor(long topologyConfigId, @NonNull ClusterDescriptor sourceCluster,
-                              @NonNull List<ClusterDescriptor> sinkClusters) {
-        this(topologyConfigId, Collections.singletonList(sourceCluster), sinkClusters);
+        remoteSinkToReplicationModel.entrySet().stream().forEach(e -> {
```

Comment:
Good catch. Thanks!
There is no need for the nested iteration at all. Since `remoteSinkToReplicationModel` is a map of cluster -> replicationModels, there is no way that a cluster can come in more than once so the nested iteration is irrelevant.